ucp/2024-03-planes-apis.md

```diff
@@ -0,0 +1,335 @@
+
+The proposal is to define separate resource types for each plane. If we were starting fresh today this is how we would do it.
+
+The best reason to change is that the current design allows nonsense. There's nothing stopping you from using an AWS plane definition and registering it as an Azure plane. The API design says that's fine. 
+
+### Design details
+
+*This is primarily an API change. Beyond API changes the intention is to keep the behavior consistent.*
+
+Currently the planes controllers are shared between all plane types, and there is a single datamodel and converter. In the proposed design we will continue to use the existing controllers, but a separate datamodel and converter per-plane-type.
+
+This is actually a simplification for features like the Radius proxy, resource groups, and tracked resources. Because these are specific to Radius they can remove a few validation steps. They currently need to protect against error cases like an non-Radius plane being registered as a Radius plane.
+
+### API design (if applicable)
+
+#### Current Design
+
+The current design for the plane resource looks like this:
+
+```typespec
+@doc("The plane resource")
+model PlaneResource is TrackedResourceRequired<PlaneResourceProperties, "planes"> {
+  @key("planeType")
+  @doc("The plane type.")
+  @segment("planes")
+  @path
+  name: string;
+}
+
+@doc("Plane kinds supported.")
+enum PlaneKind {
+  @doc("UCP Native Plane")
+  UCPNative,
+
+  @doc("Azure Plane")
+  Azure,
+
+  @doc("AWS Plane")
+  AWS,
+}
+
+@doc("The Plane properties.")
+model PlaneResourceProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  @doc("The kind of plane")
+  kind: PlaneKind;
+
+  @doc("URL to forward requests to for non UCP Native Plane")
+  url?: string;
+
+  @doc("Resource Providers for UCP Native Plane")
+  resourceProviders?: Record<string>;
+}
+```
+
+Some examples:
+
+```json
+{
+  "id": "/planes/aws/aws",
+  "name": "aws",
+  "properties": {
+    "kind": "AWS"
+  }
+}
+```
+
+
+```json
+{
+  "id": "/planes/azure/azurecloud",
+  "name": "azurecloud",
+  "properties": {
+    "kind": "Azure",
+    "url": "https://management.azure.com"
+  }
+}
+```
+
+```json
+{
+  "id": "/planes/radius/local",
+  "name": "local",
+  "properties": {
+    "kind": "UCPNative",
+    "resourceProvider": {
+      "Applications.Core": "http://applicaitons-rp:8080""
+    }
+  }
+}
+```
+
+This design leverages discriminated unions. The `type` and `location` fields are missing from the response due to bugs.
+
+#### Proposed Design
+
+There are 4 parts to the API new design, because we're separating out one type into 4.
+
+- A "generic" plane resource that's used with `/planes` (list all planes of all types)
+- AWS plane: `System.Planes/aws`
```

Comment:
How do you envision specifying cloud partition in this? I believe last part of current id allowed providing cloud partition `/planes/aws/aws-cn` for example

Reply:
This is a resource type, not a resource ID. The resource type for a scope doesn't appear in the URL. The resource IDs/URLs will be unchanged from this proposal, which is why it wasn't discussed 👍